resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/MountedDependencyManagerImpl.scala

```diff
@@ -0,0 +1,320 @@
+   * Upload submitter-local files to the resource staging server.
+   * @return The resource ID and secret to use to retrieve these files.
+   */
+  def uploadFiles(): StagedResourceIdentifier
+
+  def configurePodToMountLocalDependencies(
+    driverContainerName: String,
+    initContainerSecret: Secret,
+    initContainerConfigMap: ConfigMap,
+    originalPodSpec: PodBuilder): PodBuilder
+
+  def buildInitContainerSecret(jarsSecret: String, filesSecret: String): Secret
+
+  def buildInitContainerConfigMap(
+    jarsResourceId: String, filesResourceId: String): ConfigMap
+
+  /**
+   * Convert the Spark jar paths from their locations on the submitter's disk to
+   * the locations they will be downloaded to on the driver's disk.
+   */
+  def resolveSparkJars(): Seq[String]
+
+  /**
+   * Convert the Spark file paths from their locations on the submitter's disk to
+   * the locations they will be downloaded to on the driver's disk.
+   */
+  def resolveSparkFiles(): Seq[String]
+}
+
+private[spark] class MountedDependencyManagerImpl(
+    kubernetesAppId: String,
+    podLabels: Map[String, String],
+    podNamespace: String,
+    stagingServerUri: String,
+    initContainerImage: String,
+    jarsDownloadPath: String,
+    filesDownloadPath: String,
+    downloadTimeoutMinutes: Long,
+    sparkJars: Seq[String],
+    sparkFiles: Seq[String],
+    stagingServiceSslOptions: SSLOptions,
+    retrofitClientFactory: RetrofitClientFactory) extends MountedDependencyManager {
+  private val OBJECT_MAPPER = new ObjectMapper().registerModule(new DefaultScalaModule)
+
+  private def localUriStringsToFiles(uris: Seq[String]): Iterable[File] = {
+    KubernetesFileUtils.getOnlySubmitterLocalFiles(uris)
+      .map(Utils.resolveURI)
+      .map(uri => new File(uri.getPath))
+  }
+  private def localJars: Iterable[File] = localUriStringsToFiles(sparkJars)
+  private def localFiles: Iterable[File] = localUriStringsToFiles(sparkFiles)
+
+  override def uploadJars(): StagedResourceIdentifier = doUpload(localJars, "uploaded-jars")
+  override def uploadFiles(): StagedResourceIdentifier = doUpload(localFiles, "uploaded-files")
+
+  private def doUpload(files: Iterable[File], fileNamePrefix: String): StagedResourceIdentifier = {
+    val filesDir = Utils.createTempDir(namePrefix = fileNamePrefix)
+    val filesTgz = new File(filesDir, s"$fileNamePrefix.tgz")
+    Utils.tryWithResource(new FileOutputStream(filesTgz)) { filesOutputStream =>
+      CompressionUtils.writeTarGzipToStream(filesOutputStream, files.map(_.getAbsolutePath))
+    }
+    // TODO provide credentials properly when the staging server monitors the Kubernetes API.
+    val kubernetesCredentialsString = OBJECT_MAPPER.writer()
+      .writeValueAsString(KubernetesCredentials(None, None, None, None))
+    val labelsAsString = OBJECT_MAPPER.writer().writeValueAsString(podLabels)
+
+    val filesRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.MULTIPART_FORM_DATA), filesTgz)
```

Comment:
can this take an `InputStream` instead of a `File`?  In that case we may not have to create the tarball on disk before submitting -- we could create the tarball and stream out to the server all in one step!